### PR TITLE
Docs: Move data types to Topics section

### LINKED ISSUE #4449: Docs: move 'datatypes' content to new topic section
This is a structure that was left over from the old documentation and needs to be removed, but first we need to check if all the content has been re-assigned in the new documentation. The files considered are:

- [ ] [bands.rst](https://aiida-core.readthedocs.io/en/v1.1.0/datatypes/bands.html)
- [ ] [functionality.rst](https://aiida-core.readthedocs.io/en/v1.1.0/datatypes/functionality.html)
- [ ] [index.rst](https://aiida-core.readthedocs.io/en/v1.1.0/datatypes/index.html)
- [ ] [kpoints.rst](https://aiida-core.readthedocs.io/en/v1.1.0/datatypes/kpoints.html)

### Proposition:

It is not clear to me what to do with this, maybe we should create a "Topics/data" section for this?

### PR DESCRIPTION
**Pull request by mbercx**
Fixes #4449 

There was still content from the previous documentation on data types that was missing from the revamped docs. Here we move this content into the "Topics - Data types" section.

### Done

- Move all the content from the source/datatypes directory to the 'Data types' topics section, removing the source/datatypes directory afterwards.

- Restructure the content to first discuss the 'core' types, which we intend to keep in aiida-core in the long term, then the 'materials science' types, which may be moved to a plugin at some point. Move the export content to the end of the file.

- Reformat the content to the current style preferences:
  * Use one sentence per line.
  * Use the proper section headers.
  * Add the references in the proper format.
  * Switch to using the `code-block` directive instead of `::`.

### TODO

The section can still do with quite a large overhaul, I think. Some items on the TODO list:

- [x] Write a short text for each data type, which can start from the "Aim" bullet point for each type.
- [x] Move the other bullet points into a table, perhaps one for the "Core" and "Materials design" types each.
- [x] Show some simple use cases (e.g. how to load a dictionary into a `Dict` node) for each of the data types.
- [ ] Test all of the `KpointsData` and `BandsData` examples. Consider removing legacy code.
- [ ] Consider moving "Export data nodes to various formats" content elsewhere:
  * Move the general text on the `export` method to the new "How to share data" section (#3998).
  * Move the data type specific information to the usage examples.
- [x] Add redirects.

### CODE REVIEW

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:20]: If this is in the "reference" section, should we say something like "follow the link to the reference page on the corresponding data class" or similar?

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:33]: I would add a warning below, saying something along the lines:

Warning! While this is convenient if e.g. you need to do simple manipulations, like multiplying two numbers, be very careful of not passing such nodes instead of the corresponding python values to libraries that perform heavy computations with them: in fact, any operation on the value would be replaced with an operation creating new AiiDA nodes, that however can be orders of magnitude slower. In this case, remember to pass the `node.value` to the mathematical function instead. One can reference #3320

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:35]: Since for most classes you can retrieve the underlying value with `.value`, but for lists and dicts it's different, it would be good to add a small paragraph on this

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:48]: date time objects are now not storable anymore. We should say it can store any JSON-serializable data.
One can use utility functions from the `datetime` module or from eg. this:
https://github.com/aiidateam/aiida-core/blob/0fb6f72333b0e553590326fc348acbee3ef0763b/aiida/common/timezone.py#L122
to convert datetimes to either strings or floats (seconds from epoch) depending on the querability needs.

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:60]: Probably here an example would be useful, and also in some below (like SinglefileData) to see what the keys are

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:94]:
```suggestion
* **What is stored in the database**: the absolute path of the remote folder (and the remote computer as a `.computer` property, not as an attribute).
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:123]:
```suggestion
* **What is stored in the database**: all atomic positions, species, kinds, periodic boundary conditions, unit cell
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:138]:
```suggestion
  Beside the coordinates, it can also optionally store velocities.
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:137]:
```suggestion
* **Aim**: store molecular trajectories (i.e. sequences of StructureData objects, where then number of atomic kinds and sites does not change over time).
  Note however that the internal data format is different than StructureData, see below.
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:205]: Should we mention as the 'best' way to install this, to use instead the correct extras in `pip install aiida-core`?

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:384]: As you already mentioned, maybe all this section should be moved to an 'exporting data' section - like in "importing", we have two types (import an AiiDA DB, and import from external sources). Also here, we might want to have two types of export: export to AiiDA export file, and export to other formats (this section for the AiiDA data types).

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:35]: I agree, but just out of curiosity: why is this different? Also, should we perhaps consider changing the `__str__` method of e.g. `List` in this case, as now this can be somewhat confusing:

```Python
In [1]: l = List(list=[1, 2])

In [2]: l
Out[2]: <List: uuid: fc6335b7-69d9-4d65-bc1a-9adb5242a69f (unstored) value: [1, 2]>

In [3]: l.value
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-aa957b8aeff4> in <module>
----> 1 l.value

AttributeError: 'List' object has no attribute 'value'
```

Comment (mbercx) [reply docs/source/topics/data_types.rst:35]: 👍 I'm combining `List` and `Dict` into a subsection called "Iterables".

Comment (mbercx) [reply docs/source/topics/data_types.rst:35]: Another question I had regarding the `List` and `Dict` classes is the fact that the user cannot simply initialize e.g. a `List` using:

```Python
l = List([1, 2, 4])
```

Instead, he or she has to specify the `list` argument with its key:

```Python
l = List(list=[1, 2, 4])
```

It's also not immediately clear from the constructor signature or docstring that you have to initialize the `List` this way:

```Python
In [1]: List?
Init signature: List(**kwargs)
Docstring:      `Data` sub class to represent a list.
Init docstring:
:param backend_entity: the backend model supporting this entity
:type backend_entity: :class:`aiida.orm.implementation.entities.BackendEntity`
File:           ~/envs/aiida-docs/code/aiida-core/aiida/orm/nodes/data/list.py
Type:           AbstractNodeMeta
Subclasses:    
```

Is there a good reason for this? I.e. why don't we change the current constructor:

```Python
def __init__(self, **kwargs):
    data = kwargs.pop('list', list())
    super().__init__(**kwargs)
    self.set_list(data)
```

to something like:


```Python
def __init__(self, input_list=None, **kwargs):
    input_list = input_list or list()
    super().__init__(**kwargs)
    self.set_list(input_list)
```

If so, we should at least add a docstring to the constructor. 😅

Comment (giovannipizzi) [reply docs/source/topics/data_types.rst:35]: On the first point: indeed, adding a .value to both List and Dict would be OK I think - do you want to open an issue, and/or ask this later at the meeting? I don't know if we also need to remove get_list(), get_dict(), maybe it's OK to keep both?

On the second point: 
- we cannot call it 'input_list' as this is backward-incompatible, we need to call it list (and of course this shadows python's list - but then one can write something like
  ```python
  import builtins
  list = list or builtins.list()
  ```
  (with a comment explaining what's going on ;-) )
- I think there was some historical reasons because, for *any* node, if you passed a kwarg `aaa`, it would have been passed to a method `set_aaa`. Probably this does not exist anymore? (pinging @sphuber) Anyway, maybe this is OK and what you suggest is OK, apart from the parameter name?

**Comment by ramirezfranciscof** [comment]
@mbercx does this still require discussion or was this settled yesterday? If it does need discussion, let's remember to bring it up during today's meeting at 11.

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:48]: 👍 I've removed it from the list here.

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:380]: So far I have moved the CLI examples to the corresponding data type section (as they are rather specific) and kept the general description of the `export()` method at the end of the file. I suppose it could be fit into the "How to share data" section?

Comment (giovannipizzi) [reply docs/source/topics/data_types.rst:380]: Yes, but I think we also discussed the option to just leave them at the bottom of the page in yesterday's meeting - I'm not sure which solution is better

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:259]: I agree that some usage examples should be collapsible, but I'm not sure if this should be the case for all usage examples. The ones I've made for the core types are not very long, so these should be ok. I'll make some examples collapsible based on their length and generality, and then we can have another look.

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:202]: You mean to install the `atomic_tools` extras via:

```
 pip install aiida-core[atomic_tools]
```

?

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst]: Thanks @giovannipizzi! I opened issue #4495 to discuss these suggestions further.

**Comment by mbercx** [comment]
Thanks @giovannipizzi for the review! Besides your comments, I've already made a lot of changes to the previous version. I think the section is already in pretty good shape, but there are still some data types missing (`CifData`, `OrbitalData` and `ProjectionData`). However, since I still need to familiarize myself with these classes and their API before I can write their docs, this can still be quite time consuming. So either we:

* Continue with the content we have now so we can merge the PR asap, and open an issue to add the other data types in a next round of documentation days.
* Keep the PR open until I've added the discussion/examples of the other data types (but I'm not sure if I can manage that this week).

Let me know what you think is best! In case we want to merge it asap, I'll also look for another reviewer since you indicated you'd be busy today.

**Comment by chrisjsewell** [comment]
Before you merge this, please ensure that `datatypes/index.rst` is redirected to a suitable place (in `docs/source/redirects.txt`)

**Comment by mbercx** [comment]
@giovannipizzi this is ready for another round of review! Also fine to merge in its current state, I'll give the whole section another pass when adding the missing data types (#4529).

File: docs/source/redirects.txt
Comment (giovannipizzi) [inline docs/source/redirects.txt:20]: Let's also redirect the other deleted files

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:105]: I think the idea is to have a newline after each full stop.

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:157]: Do we need this line?

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:294]:
```suggestion
Since the :py:class:`~aiida.orm.nodes.data.folder.FolderData` node is simply a collection of files, it simply stores these files in the repository.
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:313]: A big warning is due here - this will open a SSH connection, so this should NOT be done in scripts and even less in workflows; it's just a utility for the user.

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:431]:
```suggestion
* ``cif`` (export to CIF format, without symmetry reduction, i.e. always storing the structure as P1 symmetry)
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:447]: We should emphasise here that, at variance with the StructureData, here coordinates are stored in an array and not in the DB.

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:467]:
```suggestion
* ``cif`` (export to CIF format, without symmetry reduction, i.e. always storing the structures as P1 symmetry)
```

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:472]: Do we want to keep a todo in the docs? Maybe better to open an issue?

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [reply docs/source/topics/data_types.rst:554]: yes

File: docs/source/topics/data_types.rst
Comment (giovannipizzi) [inline docs/source/topics/data_types.rst:708]:
```suggestion
      bands_spins = [bands, bands-0.3] # to distinguish the bands of different spins we subtract 0.3 from the second band structure
```

**Review by giovannipizzi — CHANGES_REQUESTED** [review]
Thanks @mbercx ! A few comments to address, then it's ok to merge with me and the rest can be addressed in a future PR

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:157]: Whoops! That's some sloppy copy-paste work, my bad! This line needs to import the `ArrayData` class, of course. Thanks for spotting it. 👍

File: docs/source/topics/data_types.rst
Comment (mbercx) [reply docs/source/topics/data_types.rst:472]: I've opened an issue on adding the missing data types to this section in #4529. This todo is just to make sure I also expand the usage section on `TrajectoryData` a bit when I work on these.

Comment (ramirezfranciscof) [reply docs/source/topics/data_types.rst:472]: I think what we were currently doing is both. So maybe add a reference to that PR in the `.. todo::` (see other sections like `internals/engine.rst` for reference) and also mention in the issue the specific change expected here.

Comment (mbercx) [reply docs/source/topics/data_types.rst:472]: 👍  Alright, thanks @ramirezfranciscof!

**Comment by giovannipizzi** [comment]
Thanks @mbercx !

**Pull request merged**
